snorkel/analysis/utils.py

```diff
@@ -12,21 +12,73 @@ def set_seed(seed: int) -> None:
     torch.manual_seed(seed)
 
 
-def probs_to_preds(probs: np.ndarray) -> np.ndarray:
+def break_ties(
+    Y_prob: np.ndarray, tie_break_policy: Optional[str] = "random"
+) -> np.ndarray:
+    """Break ties among probabilistic labels according to given policy.
+
+    Policies to break ties include:
+    "abstain": return an abstain vote (0)
+    "true-random": randomly choose among the tied options
+    "random": randomly choose among tied option using deterministic hash
+
+    NOTE: if tie_break_policy="true-random", repeated runs may have slightly different results due to difference in broken ties
+
+    Parameters
+    ----------
+    Y_prob
+        An [n,k] array of probabilistic labels
+    tie_break_policy
+        Policy to break ties, by default 'random'
+
+    Returns
+    -------
+    np.ndarray
+        An [n,1] array of integer labels
+    """
+
+    n, k = Y_prob.shape
+    Y_pred = np.zeros(n)
+    diffs = np.abs(Y_prob - Y_prob.max(axis=1).reshape(-1, 1))
+
+    TOL = 1e-5
+    for i in range(n):
+        max_idxs = np.where(diffs[i, :] < TOL)[0]
+        if len(max_idxs) == 1:
+            Y_pred[i] = max_idxs[0] + 1
+        # Deal with "tie votes" according to the specified policy
+        elif tie_break_policy == "random":
+            Y_pred[i] = max_idxs[i % len(max_idxs)] + 1
+        elif tie_break_policy == "true-random":
+            Y_pred[i] = np.random.choice(max_idxs) + 1
+        elif tie_break_policy == "abstain":
+            Y_pred[i] = 0
+        else:
+            raise ValueError(
+                f"tie_break_policy={tie_break_policy} policy not recognized."
+            )
+    return Y_pred
+
+
+def probs_to_preds(
```

Comment:
Do probs_to_preds need to be public functions?

Reply:
As in, do we want them to show up in the API docs?

Reply:
I think so. It's a frequent enough operation, especially since a lot of other models are likely to output probabilities in certain circumstances, and it provides an easy way for people to use the same tie-breaking policies as us so that it's apples to apples. (Also, with our current 1-indexed convention, this method is what performs the +1 offset that makes an [n, 2] matrix of probs become 1s and 2s instead of 0s and 1s).